File: apps/mobile/test/protocol_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:ccpocket/models/messages.dart';
import 'package:ccpocket/models/protocol_version.dart';
import 'package:flutter_test/flutter_test.dart';

Map<String, dynamic> fixture(String name) {
  final contents = File('../../test/fixtures/protocol/v1/$name.json')
      .readAsStringSync();
  return jsonDecode(contents) as Map<String, dynamic>;
}

void main() {
  group('protocol v1 contract fixtures', () {
    for (final name in ['legacy-session-list', 'current-session-list']) {
      test('accepts $name', () {
        final json = fixture(name);
        final compatibility = ProtocolCompatibility.fromBridgeJson(json);
        final message = ServerMessage.fromJson(json);

        expect(compatibility.isCompatible, isTrue);
        expect(compatibility.selectedVersion, 1);
        expect(message, isA<SessionListMessage>());
      });
    }

    for (final name in [
      'legacy-client-capabilities',
      'current-client-capabilities',
    ]) {
      test('keeps $name as a frozen client fixture', () {
        final json = fixture(name);

        expect(json['type'], 'client_capabilities');
        expect(json['supportedServerMessages'], isA<List<dynamic>>());
      });
    }
  });
}

File: packages/bridge/src/protocol-contract.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { parseClientMessage } from "./parser.js";
import {
  clientProtocolRange,
  negotiateProtocolVersion,
} from "./protocol-version.js";

function fixture(name: string): string {
  return readFileSync(
    new URL(`../../../test/fixtures/protocol/v1/${name}.json`, import.meta.url),
    "utf8",
  );
}

describe("protocol v1 contract fixtures", () => {
  it.each(["legacy-client-capabilities", "current-client-capabilities"])(
    "accepts %s",
    (name) => {
      const message = parseClientMessage(fixture(name));
      expect(message?.type).toBe("client_capabilities");
      if (message?.type !== "client_capabilities") return;

      expect(
        negotiateProtocolVersion(clientProtocolRange(message)),
      ).toBe(1);
    },
  );

  it.each(["legacy-session-list", "current-session-list"])(
    "keeps %s as a frozen server fixture",
    (name) => {
      const message = JSON.parse(fixture(name)) as Record<string, unknown>;
      expect(message.type).toBe("session_list");
      expect(message.sessions).toEqual([]);
    },
  );
});
